tsnet/packet_filter_test.go

```diff
@@ -0,0 +1,234 @@
+				{src: "2.2.2.2", dst: "1.1.1.2", port: 22, want: filter.Drop}, // different dst
+			},
+		},
+		{
+			name: "capmap_based_peers_changed",
+			initialNetMap: &tailcfg.MapResponse{
+				Node: &tailcfg.Node{
+					Addresses: []netip.Prefix{netip.MustParsePrefix("1.1.1.1/32")},
+					CapMap:    tailcfg.NodeCapMap{"X-sigil": nil},
+				},
+				PacketFilter: []tailcfg.FilterRule{{
+					SrcIPs: []string{"cap:label-1"},
+					DstPorts: []tailcfg.NetPortRange{{
+						IP: "1.1.1.1/32",
+						Ports: tailcfg.PortRange{
+							First: 22,
+							Last:  22,
+						},
+					}},
+					IPProto: []int{int(ipproto.TCP)},
+				}},
+			},
+			initialWaitTest: func(nm *netmap.NetworkMap) bool {
+				return nm.SelfNode.HasCap("X-sigil")
+			},
+			incrementalNetMap: &tailcfg.MapResponse{
+				PeersChanged: []*tailcfg.Node{{
+					ID:        2,
+					Name:      "foo",
+					Key:       key,
+					Addresses: []netip.Prefix{netip.MustParsePrefix("2.2.2.2/32")},
+					CapMap:    tailcfg.NodeCapMap{"label-1": nil},
+				}},
+			},
+			incrementalWaitTest: func(nm *netmap.NetworkMap) bool {
+				return len(nm.Peers) > 0
+			},
+			checks: []check{
+				{src: "2.2.2.2", dst: "1.1.1.1", port: 22, want: filter.Accept},
+				{src: "2.2.2.2", dst: "1.1.1.1", port: 23, want: filter.Drop}, // different port
+				{src: "3.3.3.3", dst: "1.1.1.1", port: 22, want: filter.Drop}, // different src
+				{src: "2.2.2.2", dst: "1.1.1.2", port: 22, want: filter.Drop}, // different dst
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			ctx, cancel := context.WithTimeout(t.Context(), 30*time.Second)
+			defer cancel()
+
+			controlURL, c := startControl(t)
+			s, _, pubKey := startServer(t, ctx, controlURL, "node")
+
+			if test.initialWaitTest(s.lb.NetMap()) {
+				t.Fatal("initialWaitTest already passes before sending initial netmap: this will be flaky")
+			}
+
+			if !c.AddRawMapResponse(pubKey, test.initialNetMap) {
+				t.Fatalf("could not send map response to %s", pubKey)
+			}
+
+			if err := waitFor(t, ctx, s, test.initialWaitTest); err != nil {
+				t.Fatalf("waitFor: %s", err)
+			}
+
+			if test.incrementalNetMap != nil {
+				if test.incrementalWaitTest == nil {
+					t.Fatal("incrementalWaitTest must be set if incrementalNetMap is set")
+				}
+
+				if test.incrementalWaitTest(s.lb.NetMap()) {
+					t.Fatal("incrementalWaitTest already passes before sending incremental netmap: this will be flaky")
+				}
+
+				if !c.AddRawMapResponse(pubKey, test.incrementalNetMap) {
+					t.Fatalf("could not send map response to %s", pubKey)
+				}
+
+				if err := waitFor(t, ctx, s, test.incrementalWaitTest); err != nil {
+					t.Fatalf("waitFor: %s", err)
+				}
+			}
+
+			pf := s.lb.GetFilterForTest()
+
+			for _, check := range test.checks {
+				got := pf.Check(netip.MustParseAddr(check.src), netip.MustParseAddr(check.dst), check.port, ipproto.TCP)
```

Comment:
could make proto a field in `check`, defaulting to `ipproto.TCP` but I don't have any particular test cases in mind yet

Reply:
Yeah, that's why I didn't do it for now. Easy to add later if we need it.